dandi/pynwb_utils.py

```diff
@@ -262,9 +262,97 @@ def _get_pynwb_metadata(path: str | Path | Readable) -> dict[str, Any]:
         # get external_file data:
         out["external_file_objects"] = _get_image_series(nwb)
 
+        # Calculate session duration for metadata
+        session_duration = _get_session_duration(nwb)
+        if session_duration is not None and out.get("session_start_time") is not None:
+            # Convert to absolute datetime by adding duration to session_start_time
+            from datetime import timedelta
+            start_time = out["session_start_time"]
+            out["session_end_time"] = start_time + timedelta(seconds=session_duration)
+
     return out
 
 
+def _get_session_duration(nwb: pynwb.NWBFile) -> float | None:
+    """Calculate the duration of a recording session from NWB file contents.
+
+    This function finds the minimum and maximum timestamps across all TimeSeries
+    and DynamicTable objects with time information, then returns the duration as
+    max - min.
+
+    Parameters
+    ----------
+    nwb: pynwb.NWBFile
+        An open NWB file object
+
+    Returns
+    -------
+    float or None
+        The session duration in seconds (max_time - min_time),
+        or None if no time information could be extracted
+    """
+    min_time = None
+    max_time = None
+
+    # Iterate through all objects in the NWB file
+    for obj in nwb.objects.values():
+        try:
+            # Handle TimeSeries objects
+            if isinstance(obj, pynwb.base.TimeSeries):
+                if obj.timestamps is not None and len(obj.timestamps) > 0:
+                    # Use first and last timestamps
+                    start_val = float(obj.timestamps[0])
+                    end_val = float(obj.timestamps[-1])
+                    if min_time is None or start_val < min_time:
+                        min_time = start_val
+                    if max_time is None or end_val > max_time:
+                        max_time = end_val
```

Comment:
too much of repeated `if`s -- I think it would be much cleaner if you just collect all `start_times, end_times: list[ float]` and then do min/max invocation on those